### ISSUE TITLE
Null-encoding converter doesn't work with PostgreSQL nested records

### ISSUE THREAD

**Issue by KuceraMartin** [jOOQ/jOOQ#17703]
### Expected behavior

If I call `withNullability(Nullability.NULL)` on the ForcedType, I would expect the conversion to be invoked every time – even if the value is null.

In the problem reproduction below, I would expect the output to be
```
(unimportant JOOQ prints...)
name: Alan Turing	age: Some(41)
name: Ada Lovelace	age: None
```

When running the reproduction, I also tried putting a breakpoint into `OptionIntConverter.from` to see where it gets called from. Here is the call stack that I got:
```
OptionIntConverter.from(Integer): Option (/Users/martin/development/jooq-option-problem/jooq-generated/src/main/scala/converters/OptionIntConverter.scala:8)
OptionIntConverter.from(Object): Object (/Users/martin/development/jooq-option-problem/jooq-generated/src/main/scala/converters/OptionIntConverter.scala:5)
ContextConverter$1.from(Object,ConverterContext): Object (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/ContextConverter.java:147)
DefaultBinding$DefaultRecordBinding.pgFromString(BindingScope,Field,String): Object (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/impl/DefaultBinding.java:4286)
DefaultBinding$DefaultRecordBinding.pgSetValue(BindingScope,Record,Field,String): void (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/impl/DefaultBinding.java:4411)
DefaultBinding$DefaultRecordBinding.lambda$pgNewRecord$13(BindingScope,List,Record): Record (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/impl/DefaultBinding.java:4403)
DefaultBinding$DefaultRecordBinding$$Lambda/0x00000008002c04e0.apply(Object): Object (Unknown Source:-1)
RecordDelegate.operate(ThrowingFunction): Record (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/impl/RecordDelegate.java:144)
DefaultBinding$DefaultRecordBinding.pgNewRecord(BindingScope,Class,AbstractRow,Object): Record (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/impl/DefaultBinding.java:4399)
DefaultBinding$DefaultRecordBinding.get0(BindingGetResultSetContext): Record (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/impl/DefaultBinding.java:4172)
DefaultBinding$DefaultRecordBinding.get0(BindingGetResultSetContext): Object (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/impl/DefaultBinding.java:4084)
DefaultBinding$InternalBinding.get(BindingGetResultSetContext): void (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/impl/DefaultBinding.java:1190)
CursorImpl$CursorRecordInitialiser.setValue(AbstractRecord,Field,int): void (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/impl/CursorImpl.java:1582)
CursorImpl$CursorRecordInitialiser.apply(AbstractRecord): AbstractRecord (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/impl/CursorImpl.java:1518)
CursorImpl$CursorRecordInitialiser.apply(Object): Object (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/impl/CursorImpl.java:1433)
RecordDelegate.operate(ThrowingFunction): Record (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/impl/RecordDelegate.java:144)
CursorImpl$CursorIterator.fetchNext(): Record (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/impl/CursorImpl.java:1390)
CursorImpl$CursorIterator.hasNext(): boolean (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/impl/CursorImpl.java:1366)
CursorImpl.fetchNext(int): Result (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/impl/CursorImpl.java:174)
AbstractCursor.fetch(int): Result (file:///Users/martin/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jooq/jooq/3.19.15/jooq-3.19.15-sources.jar!/org/jooq/impl/AbstractCursor.java:177)
```
What I found interesting is the frame `DefaultBinding:4286`:
```java
else if (type == Integer.class)
    return converter.from((T) Integer.valueOf(string), ctx.converterContext());
```
If we scroll up through the different cases of the else-ifs, we see that on the very top there is:
```java
if (string == null)
    return null;
```
In other words, if the value is null, it will indeed return null directly, whereas in other cases, the converter is invoked. 

### Actual behavior

The conversion is only invoked if the value is not null. If the value is null, the conversion is skipped entirely, and the value remains null.

In the problem reproduction below, the actual output is:
```
(unimportant JOOQ prints...)
name: Alan Turing	age: Some(41)
name: Ada Lovelace	age: null
```

### Steps to reproduce the problem

1. clone from https://github.com/KuceraMartin/jooq-option-problem
2. run `sbt generateJooq`
3. run `sbt app/run`

### jOOQ Version

3.19.15

### Database product and version

PostgreSQL 17

**Comment by lukaseder** [comment]
Thanks a lot for your report and for your reproducer. I'll have a look at it right away.

**Comment by lukaseder** [comment]
I can't get anything to run in this reproducer (I won't install sbt for this). I'll try to reproduce the issue in an independent environment, e.g. in our template for reproducers here: https://github.com/jOOQ/jOOQ-mcve

**Comment by lukaseder** [comment]
The problem is likely caused by the fact that you're projecting the nested table like this:

```scala
db.select(EMPLOYEE).from(EMPLOYEE)
```

Which nests the table into a nested record, see:
https://www.jooq.org/doc/latest/manual/sql-building/sql-statements/select-statement/select-clause/select-clause-table/

I can reproduce the issue that way. For the time being, for such simple queries, best avoid nesting records and just write:

```scala
db.selectFrom(EMPLOYEE)
```

Or something equivalent.

**Comment by lukaseder** [comment]
Here's a simpler reproducer, in Java:

```java
class OptionIntConverter extends AbstractConverter<Integer, Optional<Integer>> {
    public OptionIntConverter() {
        super(Integer.class, (Class) Optional.class);
    }

    @Override
    public Optional<Integer> from(Integer databaseObject) {
        return Optional.ofNullable(databaseObject);
    }

    @Override
    public Integer to(Optional<Integer> userObject) {
        return userObject.orElse(null);
    }
};

println(
ctx.select(row(field("cast(1 as integer)", INTEGER).convert(new OptionIntConverter())))
    .fetch()
);

println(
ctx.select(row(field("cast(null as integer)", INTEGER).convert(new OptionIntConverter())))
    .fetch()
);
```

This prints:

```
+-------------+
|nested       |
+-------------+
|(Optional[1])|
+-------------+
+--------+
|nested  |
+--------+
|({null})|
+--------+
```

**Comment by lukaseder** [comment]
> If we scroll up through the different cases of the else-ifs, we see that on the very top there is:

Great analysis. This is indeed the problem, and it appears we can just return `converter.from(null, ctx.converterContext())`

**Comment by lukaseder** [comment]
Fixed in versions:

- 3.20.0
- 3.19.16 (https://github.com/jOOQ/jOOQ/issues/17718)
- 3.18.23 (https://github.com/jOOQ/jOOQ/issues/17719)
- 3.17.32 (https://github.com/jOOQ/jOOQ/issues/17720)

**Comment by KuceraMartin** [comment]
Amazing, thanks!

**Comment by schmidti159** [comment]
Hi,
this fix caused some problems in our codebase when we upgraded to version 3.19.16.

We used this construct in some places to convert a string in the DB to an enum in the DTO:
```java
relatedLineItem.FINANCING_PRODUCT.convertFrom(FinancingProduct::valueOf))
```

This will now throw a NullPointerException, if the FINANCING_PRODUCT column is null, whereas before it just worked, because the converter was not called with a null value.

I replaced it with this, but maybe there is a better solution to provide a converter that does not support null values?
```java
relatedLineItem.FINANCING_PRODUCT
                        .convertFrom(product -> Optional.ofNullable(product)
                                .map(FinancingProduct::valueOf)
                                .orElse(null)))
```

**Comment by lukaseder** [comment]
> I'm not sure yet if additional overloads are worth it.

I don't think it's worth it. For consistency reasons, we'd have to add:

- `convertNullable(Class<U>, Function<T, U>, Function<U, T>)`
- `convertFromNullable(Class<U>, Function<T, U>)`
- `convertFromNullable(Function<T, U>)`
- `convertToNullable(Class<U>, Function<U, T>)`
- `convertToNullable(Function<U, T>)`

And that's just overkill.
